logged-evaluation: add IN_LORRI_SHELL

cole-h opened the pull request:
This will allow users to differentiate between a `nix-shell` invocation
and `lorri shell` invocation. One potential usecase is customizing the
prompt based on whether or not one is in a lorri shell.

<!--
Thank you for your contribution!

If this is the first time you are contributing to lorri, please take a look at:

https://github.com/target/lorri/blob/master/CONTRIBUTING.md
-->

<!-- Please replace ISSUE by the issue number this pull request addresses. -->
Closes #372.

## Overview

<!--
Explain the approach you took to resolving the issue and provide necessary context.

There is no need to go into a lot of detail here: instead, try to make each commit self-explanatory
and include good commit messages.

See https://github.com/target/lorri/blob/master/CONTRIBUTING.md for more on how to structure a pull request.
-->

* I added `IN_LORRI_SHELL` right below `IN_NIX_SHELL` in `logged-evaluation.nix`.
  * I opted not to implement the suggested nested check because, when combined with an `.envrc`, `lorri shell` would always be disallowed unless `IN_LORRI_SHELL` was unset. This can be easily added back in, however; let me know.
* ~~I fixed the warnings-turned-errors that cropped up when running the `ci` script in a `nix-shell`. I can split this out to another PR, if you'd prefer.~~ I've moved this commit to #396, which also addresses the GitHub Actions CI failures.

## Checklist

<!-- This checklist is here to help you and your reviewers, so feel free to edit it as appropriate,
e.g. bugfixes don't usually require a documentation change. -->

- [ ] Updated the documentation (code documentation, command help, ...)
- [x] Tested the change (unit or integration tests)
- [x] Amended the changelog in `release.nix` (see `release.nix` for instructions)

curiousleo commented on tests/integration/bug97_varmap_leak.rs:
```diff
@@ -32,6 +32,7 @@ fn bug97_varmap_leak() {
         "origExtraClosure",
         // Nix-set variables
         "IN_NIX_SHELL",
+        "IN_LORRI_SHELL",
```
Nit: this is listed under "Nix-set variables" although it is really set by lorri. I suggest creating a new "section" here, perhaps "lorri-set variables".

cole-h replied:
I noticed this too haha. I'll add another section along with the integration test.

cole-h commented:
OK, I added an integration test and a new section for `bug97`. I couldn't find a good place to put it (since it doesn't really involve the `.envrc`, I was hesitant to put it in `envrc.rs`), so I moved all the stuff present in `direnv.rs` into `direnvtestcase.rs` and repurposed `direnv.rs` into a file for tests like mine (I really didn't want to add another file for something so trivial, so I just took over an existing one).

cole-h commented:
Great idea! I've done just that.

Profpatsch commented on release.nix:
```diff
@@ -5,6 +5,13 @@
+      {
+        version = 548;
+        changes = ''
+          Add $IN_LORRI_SHELL environment variable to allow differentiation
+          between `nix-shell` and `lorri shell`.
```
We should mention that it carries the path to the evaluated shell file as value.

Profpatsch commented on src/logged-evaluation.nix:
```diff
@@ -129,9 +129,14 @@ let
-
             export IN_NIX_SHELL=impure
 
+            ${if shellSrc == null then "" else ''
+              # Export IN_LORRI_SHELL to differentiate between `lorri shell` and
+              # `nix-shell`
+              export IN_LORRI_SHELL=${shellSrc}
+            ''}
```
The `null` conditional looks strange to me; @curiousleo do we still require this switch for the services? I think we only ever had a WIP for services, do we want to keep them around in the codebase?

cole-h replied:
If `shellSrc ? null` turns out to be unnecessary (pending @curiousleo's response), should I just remove `? null` from the argset?

curiousleo replied:
I suggest we merge #405 first and rebase this PR on top of it. I've filed #404 for follow-up work to remove complexity due to the incomplete `lorri services` implementation.

Profpatsch commented:
I merged #405, so if you rebase we can just include it unconditionally.

Profpatsch commented on src/logged-evaluation.nix:
```diff
@@ -122,9 +122,12 @@ let
 
+            # Export IN_LORRI_SHELL to differentiate between `lorri shell` and
+            # `nix-shell`
+            export IN_LORRI_SHELL=${src}
```
This needs to be escaped via `lib.escapeShellArg`. I think it’s gonna be a lot of work importing `lib` into this file, so let’s just inline the definition from `nixpkgs/lib/strings.nix` into this file:

```
  escapeShellArg = arg: "'${builtins.replaceStrings ["'"] ["'\\''"] (toString arg)}'";
```

cole-h replied:
What about spaces? Should we worry about those? As it stands, having a folder `asd fg'h` will break e.g. `stat $IN_LORRI_SHELL`.

Profpatsch replied:
`escapeShellArg` will do the right thing for both spaces and single quotes. It wraps everything in single quotes, taking care of spaces, and then it leaves the single quote every time it encounters one, to prevent escaping issues.

`ab c'de` -> `'ab c'\''de'`

Bash implicitly concatenates two strings if they have no spaces between each other, so both single-quoted strings are merged into one string.

Profpatsch commented:
Thanks!
